Add test plan for zero downtime scaling of NGF (#1174)
nginxinc/nginx-gateway-fabric: tests/zero-downtime-scaling/zero-downtime-scaling.md

The rest of the test environment is the same for both cases:

- Tester VMs:
  - Configuration:
    - Debian
    - Install packages: wrk, curl, gnuplot.
  - Location - same zone as the Kubernetes cluster.
  - First VM - for HTTP traffic.
  - Second VM - for sending HTTPs traffic.
- NGF
  - Deployment with 1 replicas (during the test we will scale this up to a max of 25 replicas).
  - Exposed via a Service with type LoadBalancer, private IP.
  - Configure delayed termination with a 40-second sleep.
  - Gateway, two listeners - HTTP and HTTPs.
  - Two apps:
    - Coffee - 3 replicas.
    - Tea - 3 replicas.
  - Two HTTPRoutes
    - Coffee (HTTP).
    - Tea (HTTPS).

Notes:

- For sending traffic, we will use both wrk and curl.
  - *wrk* will generate a lot of traffic continuously, and it will have a high chance of catching of any
    (however small) periods of downtime.
  - *curl* will generate 1 request every 0.1s. While it might not catch small periods of downtime, it will
    give us timeline of failed request for big periods of downtime, which wrk doesn't do.

## Tests

Run the following tests on both the 10 and 25 node clusters.

### Setup

- Create a cluster.

> - Deploy NGF, the latest stable version with 1 replicas.

Reviewer: since we're not tied to any specific version
```suggestion
- Deploy NGF with 1 replicas.
- ```

>   - If testing with the 25 node cluster, add a pod affinity rule.

Reviewer: ```suggestion
  - If testing with the 25 node cluster, add a pod affinity rule to ensure no more than two pods will run on the same node.
```

- Expose NGF via a Service Load Balancer, internal (only accessible within the Google Cloud region) by
  adding `networking.gke.io/load-balancer-type: "Internal"` annotation to the Service.
- Deploy backend apps.
- Configure Gateway.
- Expose apps via HTTPRoutes.
- Check statuses of the Gateway and HTTPRoutes.
- Start sending traffic using wrk from tester VMs: HTTP and HTTPs.
- Check that NGINX access logs look good (all responses 200).
- Check that there are no errors in NGINX errors logs and NGF error logs.

### Scale Gradually

Do the following until there are 25 NGF Pods running:

- Scale up by 1 Pod.

> - Wait for the new Pod to become ready.

Reviewer: I wonder if there is an easy way to check that the new pod starts receiving traffic, so we know that it happens.
Checking NGINX access logs is an option.

Author: Access logs might be the easiest way. We could also check the nginx stub_status metrics but that's not any easier than logs.

Author: This wasn't feasible during testing. I did confirm that all of the Pods handled traffic. The distribution of access logs across Pods (visible in the GKE log explorer) was what I expected.

- Update the Gateway resource - add one listener - and make sure NGF processed the update (it updates the status of the
  Gateway resource accordingly).

> The, do the following until there are 1 NGF Pods running:

Reviewer: ```suggestion
Then, do the following until there are 1 NGF Pods running:
```

- Scale down by 1 Pod.
- Wait for the old Pod to terminate.
- Update the Gateway resource - remove one listener - and make sure NGF processed the update (it updates the status of the
  Gateway resource accordingly).

Now, [analyze](#analyze) the results.

### Scale Abruptly

Scale up:

- Scale up to 25 Pods.
- Wait for the new Pods to become ready.
- Update the Gateway resource - add one listener - and make sure NGF processed the update (it updates the status of the
   Gateway resource accordingly).

Scale down:

> - Scale up to 25 Pods.

Reviewer: ```suggestion
- Scale down to 1 Pod.
- ```

Author: whoops

> - Wait for the new Pods to become ready.

Reviewer: ```suggestion
- Wait for the old Pods to terminate.
```

- Update the Gateway resource - remove one listener - and make sure NGF processed the update (it updates the status of the
  Gateway resource accordingly).

Now, [analyze](#analyze) the results.

### Analyze

After each test analyze and record the following:

- Tester VMs:
  - Analyze the output of wrk commands for errors and latencies.
  - Create graphs from curl output (TODO: add instructions).
- If scaling down, check the old Pods logs in Google Monitoring:
  - NGINX Access logs - we expect only 200 responses.
    Google Monitoring query:

      ```text
      severity=INFO
      "GET" "HTTP/1.1" -"200"
      ```

  - NGINX Error logs - we expect no errors or warnings:
    Google Monitoring query:

      ```text
      severity=ERROR
      SEARCH("`[warn]`") OR SEARCH("`[error]`")
      ```

  - NGF logs - we expect no errors.

>   - Specifically look at the NGF logs before it exited, to make sure all components shutdown correctly.

Reviewer: since there will be 25 pods, maybe there is a way to analyze NGF logs using queries only rather than inspecting each pod logs individually

Author: Good idea. I'm sure there's a way to do that with GKE

The scaling tests will be performed on two different Kubernetes clusters:

- A Kubernetes cluster with 10 nodes on GKE:
  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
  - Enabled GKE logging.
- A Kubernetes cluster with 25 nodes on GKE:
  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
  - Enabled GKE logging.

The 10 Node cluster will cover the case where multiple NGF Pods are running on the same Node, whereas the 25 Node
cluster will test 1 NGF Pod per Node.

The rest of the test environment is the same for both cases:

- Tester VMs:
  - Configuration:
    - Debian
    - Install packages: wrk, curl, gnuplot.
  - Location - same zone as the Kubernetes cluster.
  - First VM - for HTTP traffic.
  - Second VM - for sending HTTPs traffic.
- NGF
  - Deployment with 1 replicas (during the test we will scale this up to a max of 25 replicas).
  - Exposed via a Service with type LoadBalancer, private IP.
  - Configure delayed termination with a 40-second sleep.
  - Gateway, two listeners - HTTP and HTTPs.
  - Two apps:
    - Coffee - 10 replicas.
    - Tea - 10 replicas.
  - Two HTTPRoutes
    - Coffee (HTTP).
    - Tea (HTTPS).

Notes:

- For sending traffic, we will use both wrk and curl.

>   - *wrk* will generate a lot of traffic continuously, and it will have a high chance of catching of any

Reviewer 2: ```suggestion
  - *wrk* will generate a lot of traffic continuously and it will have a high chance of catching any
```

    (however small) periods of downtime.
  - *curl* will generate 1 request every 0.1s. While it might not catch small periods of downtime, it will give us

>     timeline of failed request for big periods of downtime, which wrk doesn't do.

Reviewer 2: ```suggestion
    a timeline of failed requests for big periods of downtime, which wrk doesn't do.
```

or

```suggestion
    a timeline of a failed request for big periods of downtime, which wrk doesn't do.
```

  When proxying Websocket or any long-lived connections, NGINX will not terminate until that connection is closed by
  either the client or the backend. This means that unless all those connections are closed by clients/backends before
  or during an upgrade (which is highly unlikely), NGINX will not terminate, which means Kubernetes will kill NGINX. As
  a result, the clients will see the connections abruptly closed and thus experience downtime.

  As a result, we *will not* use any long-live connections in this test, because NGF cannot support zero-downtime
  upgrades in this case.

- Scaling to over 25 replicas.

## Test Environment

The scaling tests will be performed on two different Kubernetes clusters:

- A Kubernetes cluster with 10 nodes on GKE:
  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
  - Enabled GKE logging.
- A Kubernetes cluster with 25 nodes on GKE:
  - Node: n2d-standard-4 (4 vCPU, 8GB memory).
  - Enabled GKE logging.

The 10 Node cluster will cover the case where multiple NGF Pods are running on the same Node, whereas the 25 Node
cluster will test 1 NGF Pod per Node.

The rest of the test environment is the same for both cases:

> - Tester VMs:

Reviewer: The scale script requires jq. Perhaps it make sense to add another machine with configured kubectl and jq to the environment?

Author: Good callout. I think I'll remove `jq` from the script since it's not required -- it just makes the output prettier. 

Running the script from a VM would be obnoxious because the script and the manifests would have to be copied to the VM.

  - Two HTTPRoutes
    - Coffee (HTTP).
    - Tea (HTTPS).

Notes:

- For sending traffic, we will use both wrk and curl.
  - *wrk* will generate a lot of traffic continuously, and it will have a high chance of catching of any
    (however small) periods of downtime.
  - *curl* will generate 1 request every 0.1s. While it might not catch small periods of downtime, it will give us
    timeline of failed request for big periods of downtime, which wrk doesn't do.

## Tests

Run the following tests on both the 10 and 25 node clusters.

### Setup

1. Create a cluster.
2. Deploy a previous latest stable version with 1 replica

   For the 25 node cluster:

    ```console
     helm install my-release oci://ghcr.io/nginxinc/charts/nginx-gateway-fabric  --create-namespace --wait -n nginx-gateway --version 0.0.0-edge --values ./values-25-node.yaml
    ```

   For the 10 node cluster:

    ```console
     helm install my-release oci://ghcr.io/nginxinc/charts/nginx-gateway-fabric  --create-namespace --wait -n nginx-gateway --version 0.0.0-edge --values ./values-10-node.yaml
    ```

3. Expose NGF via a Service Load Balancer, internal (only accessible within the Google Cloud region) by adding

>    `networking.gke.io/load-balancer-type: "Internal"` annotation to the Service.

Reviewer 3: This is done via the previous helm install command, right?

Author: yep, thanks for catching. I'll remove
